### PR TITLE
DPCPP cooperative group

### PR DESCRIPTION
**Pull request by yhmtsai**
This PR adds the cooperative group in dpcpp to keep the same interface as cuda/hip.
Note. the subgroup seems to be warp in cuda not subwarp.
Reference: https://intel.github.io/llvm-docs/cuda/opencl-subgroup-vs-cuda-crosslane-op.html

~~This PR is WIP but I would like to bring it to review.
to check the config selection can be acceptable from ginkgo
for config selection, please check the cg_shuffle_config_call in dpcpp/test/components/cooperative_groups_kernels.dp.cpp~~

Summary:
- dim3 and sycl_nd_range: a cuda-like usage for sycl range and nd_range with tests
- `helper` gives default implementation macro for simple kernel cases (no explicit template parameter and 1d block)
- `__WG_BOUND__` gives something like `__launch_bound__` but it needs the 3d information not the product
`__WG_BOUND_CONFIG__` can use ConfigSet for easy unpack
- cooperative group implementation and set the test result individually
- another selection for config (it allows bool, int, size_type template by roughly go through all kernel template)
- update format_header such that it can handle the generated dpcpp file (the script is not yet here)
- ConfigSet: @tcojean provides a general way to embed information into encoded integer

TODO:
~~also set the cooperative group test result individually in cuda/hip?~~
I decide to move another PR to handle the cuda/hip cooperative group because it contains another subwarp test

### CODE REVIEW

**Comment by tcojean** [comment]
Thanks Mike for creating the PR. I'm trying to see what I can do with a templated configuration type as possible improvement, I'll tell you whether it works.

File: dpcpp/components/cooperative_groups.dp.hpp [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+    /**
+     * Returns a bitmask containing the value of the given predicate
+     * for all threads in the group.
+     * This means that the ith bit is equal to the predicate of the
+     * thread with thread_rank() == i in the group.
+     * Note that the whole group needs to execute the same operation.
+     */
+    __dpct_inline__ mask_type ballot(int predicate) const noexcept
+    {
+        return (predicate != 0) ? mask_type(1) << data_.rank : mask_type(0);
```
Comment (upsj) [inline dpcpp/components/cooperative_groups.dp.hpp:315]: data_.rank should be 0 here, right?

File: dpcpp/components/cooperative_groups.dp.hpp [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+        return this->ShflOp(var, selector);                                    \
+    }                                                                          \
+    static_assert(true,                                                        \
+                  "This assert is used to counter the false positive extra "   \
+                  "semi-colon warnings")
+
+    GKO_BIND_SHFL(shfl, shuffle);
+    GKO_BIND_SHFL(shfl_xor, shuffle_xor);
+
+    // the shfl_up of out-of-range value gives undetermined behavior, we
```
Comment (upsj) [inline dpcpp/components/cooperative_groups.dp.hpp:207]: What is the wording here? Does it give *undefined behavior*, or return *undefined* or *implementation-defined* values? That might be interesting for algorithms where we cause this overflow on purpose because we discard the values anyways.

File: dpcpp/components/cooperative_groups.dp.hpp [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+#endif
+        data_.rank = this->get_local_id();
+    }
+
+
+    __dpct_inline__ unsigned thread_rank() const noexcept { return data_.rank; }
+
+    __dpct_inline__ unsigned size() const noexcept { return Size; }
+
+    __dpct_inline__ void sync() const noexcept {}
```
Comment (upsj) [inline dpcpp/components/cooperative_groups.dp.hpp:191]: Maybe we need a memory fence here?

File: dpcpp/test/utils.hpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+#include <ginkgo/core/base/executor.hpp>
+
+
+namespace {
+
+
+// prevent device reset after each test
+auto no_reset_exec =
+    gko::DpcppExecutor::create(0, gko::ReferenceExecutor::create());
```
Comment (upsj) [inline dpcpp/test/utils.hpp:48]: Is this necessary at all? Does SYCL destroy its state when we destroy the executor? Do we need to do something like cudaDeviceReset() to make profiling/sanitizers work?

**Review by upsj — COMMENTED** [review]
I didn't go into the config right now, but the rest looks good

File: dpcpp/components/cooperative_groups.dp.hpp (hunk 2 above)
Comment (yhmtsai) [reply dpcpp/components/cooperative_groups.dp.hpp:207]: it is undefined behavior. we do not use it in algorithm actually, but we use it to pass cooperative group test

File: dpcpp/components/cooperative_groups.dp.hpp (hunk 1 above)
Comment (yhmtsai) [reply dpcpp/components/cooperative_groups.dp.hpp:315]: Yes

**Comment by yhmtsai** [comment]
test_all_subgroup will run 4, 8, 16 for cpu and 8, 16, 32 for intel gpu
search dpcpp/test/components/cooperative_groups_kernels on [dgpu-job](https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/1232217732), or [cpu-job](https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/1232217727)

**Review by Slaedr — APPROVED** [review]
Overall, this looks good!

The main point of interest for me is the choice of 32 for the warp size. Could you document somewhere, maybe just in this PR, why you chose 32? Will shuffles be expected to work for 32 work-items independent of vector architecture? Can we get away with a smaller warp size? (A smaller warp size, if it still fully utilizes the vector units, can enable more flexible parallelism.)

Apart from that I have a few other points.

File: dpcpp/base/config.hpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+struct config {
+    /**
+     * The type containing a bitmask over all lanes of a warp.
+     */
+    using lane_mask_type = uint64;
+
+    /**
+     * The number of threads within a DPCPP warp.
+     */
+    static constexpr uint32 warp_size = 32;
```
[the remark this answers is not in the document]
Comment (tcojean) [reply dpcpp/base/config.hpp:55]: That's a good point. I think that's just a leftover from the previous implementation? Because he now uses this configuration integer (which will change in structure later), the `warp_size` is not provided by the `config.hpp` but by the hardware capacities and the config template parameter given to kernels (subgroup size). I believe the first one to match will be used.

File: dpcpp/test/utils.hpp (hunk 4 above)
Comment (yhmtsai) [reply dpcpp/test/utils.hpp:48]: I thought it is mainly to reduce the test time?

File: dpcpp/components/cooperative_groups.dp.hpp [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+    GKO_BIND_SHFL(shfl_xor, shuffle_xor);
+
+    // the shfl_up of out-of-range value gives undetermined behavior, we
+    // manually set it as the original value such that give the same result as
+    // cuda/hip.
+    template <typename ValueType, typename SelectorType>
+    __dpct_inline__ ValueType shfl_up(ValueType var,
+                                      SelectorType selector) const noexcept
+    {
+        const auto result = this->shuffle_up(var, selector);
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/components/cooperative_groups.dp.hpp:214]: I think it needs `this->` to use the generated subgroup information

File: dpcpp/components/cooperative_groups.dp.hpp [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+    __dpct_inline__ unsigned size() const noexcept { return Size; }
+
+    __dpct_inline__ void sync() const noexcept {}
+
+#define GKO_BIND_SHFL(ShflOpName, ShflOp)                                      \
+    template <typename ValueType, typename SelectorType>                       \
+    __dpct_inline__ ValueType ShflOpName(ValueType var, SelectorType selector) \
+        const noexcept                                                         \
+    {                                                                          \
+        return this->ShflOp(var, selector);                                    \
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/components/cooperative_groups.dp.hpp:198]: When I call the member function, I will use `this->`

File: dpcpp/components/cooperative_groups.dp.hpp [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+
+
+}  // namespace group
+}  // namespace dpcpp
+}  // namespace kernels
+}  // namespace gko
+
+
+// Enable group can directly use group function
+__SYCL_INLINE_NAMESPACE(cl)
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/components/cooperative_groups.dp.hpp:481]: for example,  we use the group algorithm the reduce in ballot.
If we do not do this way, when we need use oneAPI implemenation on gko's subgroup, we need to add static_cast to subgroup type.
I agree with you, they might change it at some point.

File: dpcpp/components/cooperative_groups.dp.hpp [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+    struct alignas(8) {
+        unsigned size;
+        unsigned rank;
+    } data_;
+};
+
+
+// specialization for 1
+template <>
+class thread_block_tile<1> {
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/components/cooperative_groups.dp.hpp:273]: it is for those kernels implemented by warp sense.
providing it such that we can use the same kernel on subgroup(1) to get the single thread implementation

File: dpcpp/components/cooperative_groups.dp.hpp [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+
+// types
+namespace detail {
+
+
+/**
+ * This is a limited implementation of the DPCPP thread_block_tile.
+ */
+template <unsigned Size>
+class thread_block_tile : public cl::sycl::ONEAPI::sub_group {
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/components/cooperative_groups.dp.hpp:168]: `sycl::` works for most of cases, which depends on `__SYCL_DISABLE_NAMESPACE_INLINE__` is activated or not.
https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/defines_elementary.hpp

File: dpcpp/components/cooperative_groups.dp.hpp (hunk 3 above)
Comment (upsj) [reply dpcpp/components/cooperative_groups.dp.hpp:191]: A memory fence is an intrinsic preventing the reordering of memory operations from before and after the fence, essentially making sure that reads after the fence see writes before the fence. I'm suggesting this because in CUDA, sync calls syncwarp, which, as far as I remember, does an implicit memory barrier. But maybe we also shouldn't use a sync for this, but rather an explicit threadfence.

File: dpcpp/test/utils.hpp (hunk 4 above)
Comment (upsj) [reply dpcpp/test/utils.hpp:48]: I think the dpcpp executor doesn't have the concept of a device reset, nor does it have a device mutex to count how many executors per device there are. So I think we can leave this out completely.

File: dpcpp/components/cooperative_groups.dp.hpp (hunk 3 above)
Comment (yhmtsai) [reply dpcpp/components/cooperative_groups.dp.hpp:191]: Yes, they already have a barrier for it, but I miss updating it.

File: dpcpp/test/utils.hpp (hunk 4 above)
Comment (yhmtsai) [reply dpcpp/test/utils.hpp:48]: you are right, we do not have the reset bahavior on dpcpp (I am not sure they have the concept or not).
I also test the performance. It only change the first test time, but other stays similar.
the first test timing change should be in original one but we do not record it.
Thus, I will delete this global executor

**Comment by yhmtsai** [comment]
I add a few default helper to reduce the internal layer code.
`GKO_ENABLE_DEFAULT_HOST_CONFIG(name, kernel)` generates the default implementation for those kernels without expilict template parameter.
`GKO_ENABLE_DEFAULT_CONFIG_CALL(name, callable, cfg, list)`
`cfg` gives how to decode the int and `list` gives the available list for selection.
Also, @tcojean provides a general way, ConfigSet, to embed the several information into one int.
It can use not only in dpcpp but also in cuda/hip thing like ellpack we have 2 (3) parameters for selection.

TODO:
- [ ] add documentation

**Comment by yhmtsai** [comment]
@ginkgo-project/reviewers this PR can be reviewed. Also welcome any name suggestion of these things

File: dpcpp/test/components/cooperative_groups_kernels.dp.cpp [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+namespace {
+
+
+using namespace gko::kernels::dpcpp;
+using KCfg = gko::ConfigSet<12, 7>;
+constexpr auto default_config_list =
+    ::gko::syn::value_list<int, KCfg::encode(64, 64), KCfg::encode(32, 32),
+                           KCfg::encode(16, 16), KCfg::encode(8, 8),
+                           KCfg::encode(4, 4)>();
```
Comment (tcojean) [inline dpcpp/test/components/cooperative_groups_kernels.dp.cpp:65]: To make sure this works properly, maybe you should add a few tests in `core/test/base/types.cpp` testing that encode/decode do what it should in multiple settings, (using only one bit, using one element of full width, using multiple elements, ...).

File: dpcpp/test/components/cooperative_groups_kernels.dp.cpp [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+};
+
+
+void test_assert(bool *success, bool partial)
+{
+    if (!partial) {
+        *success = false;
+    }
+}
+
```
Comment (tcojean) [inline dpcpp/test/components/cooperative_groups_kernels.dp.cpp:120]: Seems unused and simple enough that this can be removed.
```suggestion
```

**Review by tcojean — APPROVED** [review]
LGTM

File: include/ginkgo/core/base/executor.hpp [hunk 13]
```diff
@@ -811,8 +811,37 @@ class Executor : public log::EnableLogging<Executor> {
...
+        /**
+         * The validate function for Config
+         */
+        bool validate(int blocksize, int warpsize)
+        {
+            bool allowed = false;
+            for (auto &i : subgroup_sizes) {
+                allowed |= (i == warpsize);
+            }
+            return allowed && (blocksize <= max_workgroup_size);
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/base/executor.hpp:824]: the protected structure is the original one.
I add a public function to get the `const exec_info`.
Do you have any comment about using function directly or structure function?

File: dpcpp/base/dim3.dp.hpp [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+namespace gko {
+namespace kernels {
+namespace dpcpp {
+
+
+/**
+ * dim3 is a cuda-like dim3 for sycl-range, which provides the same ordering as
+ * cuda and gets the sycl-range in reverse ordering.
+ */
+struct dim3 {
```
Comment (upsj) [inline dpcpp/base/dim3.dp.hpp:49]: We are building a lot of complexity here (also the config selection dim3 integration), so I want to ask: Why do we need this additional wrapper? Can't we use SYCL primitives directly?

File: dpcpp/components/cooperative_groups.dp.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (upsj) [reply dpcpp/components/cooperative_groups.dp.hpp:198]: member functions are sometimes not found in a templated member function unless you use `this->`, which is also why all of our templated tests need to use it.

**Review by upsj — COMMENTED** [review]
Looks mostly good to me, I would like to see if we can remove a few CUDA-isms though, since they only relate to kernel launch, not the actual kernel implementation.

File: dpcpp/base/dim3.dp.hpp (hunk 14 above)
Comment (yhmtsai) [reply dpcpp/base/dim3.dp.hpp:49]: Yes, we can use the SYCL native `sycl::range<3>`
but it will gives different view on kernel launch and build the sycl::range
for example, we use `kernel<<<32, 32>>>` and sycl will need to use
`kernel(sycl::nd_range<3>(sycl::range<3>(1, 1, 32) * sycl::range<3>(1, 1, 32), sycl::range<3>(1, 1, 32)))`
but with dim3, we can still use `kernel(32, 32)`
and in the beginning, I would like to reduce the difference from cuda to dpcpp

File: dpcpp/base/dpct.hpp [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+
+// This is partial extraction from dpct/dpct.hpp of Intel
+#if defined(_MSC_VER)
+#define __dpct_align__(n) __declspec(align(n))
+#define __dpct_inline__ __forceinline
+#else
+#define __dpct_align__(n) __attribute__((aligned(n)))
+#define __dpct_inline__ __inline__ __attribute__((always_inline))
+#endif
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/base/dpct.hpp:47]: this is due to the issues from dpct before. for example, including dpct file gives error or their atomic_add has some issue, so we also have atomic_add implementation for real number in another pr

File: dpcpp/base/dim3.dp.hpp (hunk 14 above)
Comment (upsj) [reply dpcpp/base/dim3.dp.hpp:49]: In which places does that actually matter? Most of the time, we are using one-dimensional kernels, except for SpMV (multiple columns), some Dense kernels (2D) and Jacobi. Wouldn't you use `sycl::ndrange<1>(...)` directly?

File: dpcpp/base/config.hpp [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+struct config {
+    /**
+     * The type containing a bitmask over all lanes of a warp.
+     */
+    using lane_mask_type = uint64;
+
+    /**
+     * The bitmask of the entire warp.
+     */
+    static constexpr auto full_lane_mask = ~zero<lane_mask_type>();
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/base/config.hpp:55]: you need to use `mask<subgroup_size, lane_mask_type>` to get the last `subgroup_size` bits activated in `lane_mask_type`

File: dpcpp/base/dim3.dp.hpp (hunk 14 above)
Comment (yhmtsai) [reply dpcpp/base/dim3.dp.hpp:49]: I think the mapping will be different. At least, it will require us to develop own dpct to avoid any conversion like threadIdx -> get_local_id(3)

Comment (yhmtsai) [reply dpcpp/base/dim3.dp.hpp:49]: and it will gives inconsistent index sense between the kernel1D/2D/3D.
`range<1>(x)` `get_local_id(0)` is x and x is contiguous as cuda
`range<2>(x, y)` `get_local_id(0)` is still x but x is not contiguous
or `range<2>(y, x)` x is contiguous but x needs `get_local_id(1)`

Comment (upsj) [reply dpcpp/base/dim3.dp.hpp:49]: But that only matters on the launch side, right? I would assume that the reason for this is that it maps more cleanly to the nested for-loop model
```
for (; i < range[0]; i++)
    for (; j < range[1]; j++)
        for (; k < range[2]; k++)
```

File: dpcpp/components/cooperative_groups.dp.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Slaedr) [reply dpcpp/components/cooperative_groups.dp.hpp:198]: Typically, that is not needed for inline definitions (in the class body rather than outside). I was just thinking of avoiding an explicit pointer indirection if we can help it. I understand it's probably fine because it should be optimized out.

File: dpcpp/base/dim3.dp.hpp (hunk 14 above)
Comment (yhmtsai) [reply dpcpp/base/dim3.dp.hpp:49]: it's not only affect launch side but the kernel index for threads

File: dpcpp/base/dim3.dp.hpp [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        : x(xval), y(yval), z(zval)
+    {}
+
+    /**
+     * reverse returns the range for sycl with correct ordering (reverse of
+     * cuda)
+     *
+     * @return sycl::range<3>
+     */
+    sycl::range<3> reverse() { return sycl::range<3>(z, y, x); }
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/base/dim3.dp.hpp]: I change it to `get_range`

File: dpcpp/test/components/cooperative_groups_kernels.dp.cpp (hunk 11 above)
Comment (yhmtsai) [reply dpcpp/test/components/cooperative_groups_kernels.dp.cpp:65]: I add several test into `core/test/base/types.cpp` and change the type of encode/decode to unsigned

**Review by tcojean — APPROVED** [review]

**Review by Slaedr — COMMENTED** [review]
Excellent work! It's pretty cool, eg. how you handle stuff like `__WG_BOUNDS__`. Guess you want to make DPC++ as close to CUDA as possible, which is what a lot of the code here seems to be doing. 

Below, I have some concerns around the `ConfigSet` stuff and other comments / suggestions.

File: include/ginkgo/core/base/types.hpp [hunk 18]
```diff
@@ -193,6 +194,147 @@ using default_precision = double;
...
+ * shift calculates the number of bits for shifting
+ *
+ * @tparam num_groups  the number of elements in array
+ * @tparam current_shift  the current position of shifting
+ *
+ * @return the number of shifting bits
+ *
+ * @note this is the last case of nested template
+ */
+template <int num_groups, int current_shift>
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/base/types.hpp:236]: I tried it before but it did not get the num_group information from array

File: include/ginkgo/core/base/types.hpp [hunk 19]
```diff
@@ -193,6 +194,147 @@ using default_precision = double;
...
+     * @param first  the current encoded information
+     * @param rest...  the rest of others information waits for encoding
+     *
+     * @return the encoded integer
+     */
+    template <size_type current_iter = 0, typename... Rest>
+    static constexpr std::enable_if_t<(current_iter < num_groups),
+                                      ConfigSetType>
+    encode(ConfigSetType first, Rest &&... rest)
+    {
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/base/types.hpp:330]: we use this for runtime and compile time, so first is not guaranteed to be compile-time arguments.
`static_assert` requires all thing compile-time known.
`assert` limits the function into runtime only.
If we need the check, we need to distinguish runtime and compile time function.
compile time will use `ConfigSet::encode<...>`, which uses static_assert
runtime will use `ConfigSet::encode(...)`, which uses assert

File: include/ginkgo/core/base/types.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (Slaedr) [reply include/ginkgo/core/base/types.hpp:236]: Oh I see. Even with the switched order? Then never mind.

File: include/ginkgo/core/base/types.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (Slaedr) [reply include/ginkgo/core/base/types.hpp:330]: Maybe a hand-written constexpr assert function? It's fine if you think it's unnecessary. Maybe you could just create an issue to record it.

File: dpcpp/components/cooperative_groups.dp.hpp [hunk 20]
```diff
@@ -0,0 +1,513 @@
...
+
+
+// types
+namespace detail {
+
+
+/**
+ * This is a limited implementation of the DPCPP thread_block_tile.
+ */
+template <unsigned Size>
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/components/cooperative_groups.dp.hpp:167]: it needs to be template parameter to match the same usage as cuda/hip and keep `.size()` is compiletime such that it can be used in `#pragma unroll` (although dpcpp does not fully support it)

File: dpcpp/test/components/cooperative_groups_kernels.dp.cpp [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    template <typename Kernel>
+    void test_all_subgroup(Kernel kernel)
+    {
+        auto subgroup_size = GetParam();
+        auto queue = dpcpp->get_queue();
+        if (gko::kernels::dpcpp::validate(queue, subgroup_size,
+                                          subgroup_size)) {
+            const auto cfg = KCfg::encode(subgroup_size, subgroup_size);
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/test/components/cooperative_groups_kernels.dp.cpp:92]: For this, I would prefer putting it in another pr if needed.

File: include/ginkgo/core/base/types.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/base/types.hpp:236]: it needs switched order to give the possibility for not explicitly setting.
I will put the my trying code later.
It is also not my expectation, so maybe I did something wrong there.

File: include/ginkgo/core/base/types.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/base/types.hpp:330]: what do you mean hand-written constexpr assert?
putting constexpr function into static_assert or write a constexpr check and throw?
the first one I tried but fail for the same reason.
the second one I need to check throw in constexpr is allowable or not.

[the remark this answers is not in the document]
Comment (Slaedr) [reply include/ginkgo/core/base/types.hpp:330]: Yes, that does seem difficult. However, it looks like throwing from constexpr functions is mostly allowed? https://tech.jocodoma.com/2019/03/07/Exceptions-in-constexpr-Functions-in-CPP/

File: dpcpp/test/components/cooperative_groups_kernels.dp.cpp [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+#include "dpcpp/base/config.hpp"
+#include "dpcpp/base/dim3.dp.hpp"
+#include "dpcpp/base/helper.hpp"
+
+
+namespace {
+
+
+using namespace gko::kernels::dpcpp;
+using KCfg = gko::ConfigSet<12, 7>;
```
[the remark this answers is not in the document]
Comment (Slaedr) [reply dpcpp/test/components/cooperative_groups_kernels.dp.cpp:62]: That appears to make sense, but I don't really see it. It seems like you want the bits array to contain the maximum number of bits needed by each position. But I guess it will work correctly for many combinations, depending on which numbers are given as parameters to the `encode` function.

[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/test/components/cooperative_groups_kernels.dp.cpp:62]: it is described in ConfigSet note. the #bit should be `log_2(max)+1`

File: core/test/base/types.cpp [hunk 23]
```diff
@@ -100,4 +105,109 @@ TEST(PrecisionReduction, ComputesCommonEncoding)
...
+    constexpr auto fullmask_u = gko::detail::mask<32>();
+    constexpr auto mask3_u64 = gko::detail::mask<3, std::uint64_t>();
+    constexpr auto fullmask_u64 = gko::detail::mask<64, std::uint64_t>();
+
+    ASSERT_EQ(mask3_u, 7u);
+    ASSERT_EQ(fullmask_u, 0xffffffffu);
+    ASSERT_EQ((std::is_same<decltype(mask3_u), const unsigned int>::value),
+              true);
+    ASSERT_EQ((std::is_same<decltype(fullmask_u), const unsigned int>::value),
+              true);
```
Comment (upsj) [inline core/test/base/types.cpp:120]:
```suggestion
    ASSERT_TRUE((std::is_same<decltype(mask3_u), const unsigned int>::value));
    ASSERT_TRUE((std::is_same<decltype(fullmask_u), const unsigned int>::value));
```
Not sure if you even need the additional parentheses then?

File: core/test/base/types.cpp [hunk 24]
```diff
@@ -100,4 +105,109 @@ TEST(PrecisionReduction, ComputesCommonEncoding)
...
+    ASSERT_EQ(mask3_u64, 7ull);
+    ASSERT_EQ(fullmask_u64, 0xffffffffffffffffull);
+    ASSERT_EQ((std::is_same<decltype(mask3_u64), const std::uint64_t>::value),
+              true);
+    ASSERT_EQ(
+        (std::is_same<decltype(fullmask_u64), const std::uint64_t>::value),
+        true);
```
Comment (upsj) [inline core/test/base/types.cpp:127]: same here and the following

File: dev_tools/scripts/format_header.sh [hunk 25]
```diff
@@ -266,9 +266,6 @@ fi
 # Arrange the remain files and give
 if [ -f "${CONTENT}" ]; then
     add_regroup
-    if [ "${HAS_HIP_RUNTIME}" = "true" ]; then
```
Comment (upsj) [inline dev_tools/scripts/format_header.sh:269]: What is the reason for these changes again? Do we no longer need it?

File: include/ginkgo/core/base/types.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/base/types.hpp:236]: Did you try adding a [deduction guide](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction#User-defined_deduction_guides)? Neat little C++17 feature that might help here.

**Review by upsj — APPROVED** [review]
LGTM - nice job! I am still not entirely happy about the `dim3` thing, since it means adding CUDA-isms to our SYCL code, but I guess it would only be used in a small number of places anyways, so I guess it will be okay.

File: include/ginkgo/core/base/types.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/base/types.hpp:330]: it works.
throw will gives a compile time error from it is not constexpr function, which means the string can not give variable information.
throw in runtime is as usual.
some related code
https://godbolt.org/z/oa5d4arMs

File: include/ginkgo/core/base/types.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/base/types.hpp:236]: I think ConfigSet is also a general solution for others kernel (not dpcpp) so we still need to stay C++14.
I put the related code: https://godbolt.org/z/oa5d4arMs
it switch the order but can not miss the num_groups

**Comment by yhmtsai** [comment]
it depends on how much effort we would like to start DPCPP in the beginning.
If we wouldn't like Cuda-isms into dpcpp, we should reimplement all not porting in the begining.
Using dpct gives the index structure and also the global range is not possible be the same as grid.
without dim3, we still need to do the same thing manually and need to switch our brand between SYCL/(HIPCUDA)

File: dev_tools/scripts/format_header.sh (hunk 25 above)
Comment (yhmtsai) [reply dev_tools/scripts/format_header.sh:269]: the behavior still exists. DPCPP gives more additional header than <hip/runtime.h>, so I move the all additional before the LICENSE into header section not just hip/runtime.h

File: include/ginkgo/core/base/types.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (Slaedr) [reply include/ginkgo/core/base/types.hpp:330]: Excellent. We can use that.

File: dpcpp/components/cooperative_groups.dp.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dpcpp/components/cooperative_groups.dp.hpp:481]: I stay this version first because it still works and it makes ginkgo cooperative group like an extension not another object.
But if they change frequently and we are hard to stay with them, we can delete them and use static_cast.

**Comment by yhmtsai** [comment]
I swap the order of shift although we do not have auto deduction there.
But if someway to achieve it or use C++17, this order of template is unchanged.
I add some comments to show the encode/decode is bijection function.
Also the ConfigSet can not be compiled or can throw the error.
- the total bits is larger than the bits of ConfigSetType
- the value is larger than the certain bit representation
- encode uses wrong number of args
- decode out-of-range position

the review update is in https://github.com/ginkgo-project/ginkgo/pull/757/commits/6ab97cd68a9b0d57f8901a68885da22ddd056a98

File: core/test/base/types.cpp (hunk 23 above)
Comment (yhmtsai) [reply core/test/base/types.cpp]: It still needs the additional parentheses such that the macro unpacks parameter correctly.

File: include/ginkgo/core/base/types.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (tcojean) [reply include/ginkgo/core/base/types.hpp:351]: I would not rely on this. This seems like UB to me. See this other setting where you get exactly the opposite behavior:
https://stackoverflow.com/questions/34280729/throw-in-constexpr-function

[the remark this answers is not in the document]
Comment (Slaedr) [reply include/ginkgo/core/base/types.hpp:351]: I saw that and my tentative conclusion was that it can be used, though I am no expert. I guess "ill-formed" != UB. In any case, if this works with all our CI toolchains, I vote to keep it because it offers some value.

[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/base/types.hpp:351]: If the constexpr hit the condition, then the compiler will add the throw into constexpr.
However, it is impossible to put the throw in core constexpr, so the compiler error is the throw is not const statement or something.

I guess if the compiler does not ignore the condition in compile time, the compile error is always happened.

For non-constexpr, it just perform as usual.

[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/base/types.hpp:351]: I try several gcc compilers on compiler explorer.
It is failed with gcc version <= 5.x no matter hit the condition or not.
I think 5.4 is a default compiler in Ubuntu 16.04 and the CUDA 9.2/10.x also allow this compiler version.

File: include/ginkgo/core/base/types.hpp [hunk 26]
```diff
@@ -193,6 +196,179 @@ using default_precision = double;
...
+ *
+ * The usage will be the following
+ * Set the method with bits Cfg = ConfigSet<b_0, b_1, ..., b_k>
+ * Encode the given infomation encoded = Cfg::encode(x_0, x_1, ..., x_k)
+ * Decode the specific position information x_t = Cfg::decode<t>(encoded)
+ * The encoded result will use 32 bits to record
+ * rrrrr0..01....1...k..k, which 1/2/.../k means the bits store the information
+ * for 1/2/.../k position and r is for rest of unused bits.
+ *
+ * Denote B_t = sum_(t+1)^(k) b_i and F(X) = Cfg::encode(x_0, ..., x_k)
```
Comment (tcojean) [inline include/ginkgo/core/base/types.hpp:277]: Can you use `$` to wrap the math portions, that should provide a better documentation style?
